libraries/chain/include/eosio/chain/chain_config_helper.hpp

```diff
@@ -0,0 +1,186 @@
+   if (!entry.is_allowed()) return s;\
+   BOOST_PP_SEQ_FOR_EACH(CASE_PACK, CLASS, MEMBERS)\
+   if constexpr (std::is_base_of<BASE, CLASS>()) {\
+      fc::raw::pack(s, eosio::chain::data_entry<BASE, eosio::chain::config_entry_validator>(entry));\
+      return s;\
+   }\
+   FC_THROW_EXCEPTION(eosio::chain::config_parse_error, "DataStream& operator<<: no such id: ${id}", ("id", entry.id));\
+}
+
+#define RANGE_PACK(CLASS, MEMBERS)\
+RANGE_PACK_DERIVED(std::false_type, CLASS, MEMBERS)
+
+//TODO: add to CASE_UNPACK check for future protocol being active
+#define CASE_UNPACK(r, CLASS, MEMBER)\
+case eosio::chain::field_id<&CLASS::MEMBER>():\
+   fc::raw::unpack(s, entry.config.MEMBER);\
+   return s;
+
+#define RANGE_UNPACK_DERIVED(BASE, CLASS, MEMBERS)\
+template<typename DataStream>\
+inline DataStream& operator>>( DataStream& s, eosio::chain::data_entry<CLASS, eosio::chain::config_entry_validator>& entry ) {\
+   EOS_ASSERT(entry.is_allowed(), eosio::chain::unsupported_feature, "config id ${id} is no allowed", ("id", entry.id));\
+   switch(entry.id){\
+   BOOST_PP_SEQ_FOR_EACH(CASE_UNPACK, CLASS, MEMBERS)\
+   }\
+   if constexpr (std::is_base_of<BASE, CLASS>()) {\
+      eosio::chain::data_entry<BASE, eosio::chain::config_entry_validator> base_entry(entry);\
+      fc::raw::unpack(s, base_entry);\
+      return s;\
+   }\
+   FC_THROW_EXCEPTION(eosio::chain::config_parse_error, "DataStream& operator<<: no such id: ${id}", ("id", entry.id));\
+}
+
+#define RANGE_UNPACK(CLASS, MEMBERS)\
+RANGE_UNPACK_DERIVED(std::false_type, CLASS, MEMBERS)
+
+#define ENUM_PAIR2(r, CLASS, ID, MEMBER, BASE)\
+   bh::make_pair(bh::size_c<OFFSETOF_IMPL(CLASS, MEMBER)>,\
+                 bh::size_c<enum_size<BASE>() + ID>),
+
+#define ENUM_PAIR_DERIVED(r, BASE_DERIVED, ID, MEMBER)\
+ENUM_PAIR2(r, BOOST_PP_SEQ_ELEM(1, BASE_DERIVED), ID, MEMBER, BOOST_PP_SEQ_ELEM(0, BASE_DERIVED))
+
+#define DEFINE_FIELD_ID(r, CLASS, ID, MEMBER)\
+template<>\
+constexpr size_t field_id<&CLASS::MEMBER>(){\
+   auto t = bh::size_c<OFFSETOF_IMPL(CLASS, MEMBER)>;\
+   auto found = bh::find(BOOST_PP_CAT(CLASS, _MAP), t);\
+   return *found;\
+}
+
+#define DEFINE_ENUM_DERIVED(BASE, DERIVED, MEMBERS)\
+constexpr auto DERIVED##_MAP = bh::make_map(\
+   BOOST_PP_SEQ_FOR_EACH_I(ENUM_PAIR_DERIVED, (BASE)(DERIVED), MEMBERS)\
+   bh::make_pair(bh::size_c<sizeof(DERIVED)>, bh::size_c<BOOST_PP_SEQ_SIZE(MEMBERS)>)\
+);\
+BOOST_PP_SEQ_FOR_EACH_I(DEFINE_FIELD_ID, DERIVED, MEMBERS)\
+template<>\
+constexpr size_t enum_size<DERIVED>(){\
+   auto t = bh::size_c<sizeof(DERIVED)>;\
+   auto found = bh::find(DERIVED##_MAP, t);\
+   return *found;\
+}
+
+#define DEFINE_ENUM(CLASS, MEMBERS)\
+DEFINE_ENUM_DERIVED(void, CLASS, MEMBERS)
+
+/**
+ * Returns field in in the structure
+ * This is enum-kind implementation.
+ * See DEFINE_FIELD_ID macro for implementation
+ */
+template<auto T>
+constexpr size_t field_id(){
+   FC_THROW_EXCEPTION(config_parse_error, "only specializations of field_id can be used");
```

Comment:
Is there a way these could be static_asserts instead, making it a compile time failure?